render/hyperscript.js

```diff
@@ -27,22 +27,18 @@ function compileSelector(selector) {
 function execSelector(state, attrs, children) {
 	var hasAttrs = false, childList, text
 	var className = attrs.className || attrs.class
+	var classAttr = hasOwn.call(attrs, "class") ? "class" : "className"
 
 	for (var key in state.attrs) {
-		if (hasOwn.call(state.attrs, key)) {
+		if (hasOwn.call(state.attrs, key) && key !== "className") {
 			attrs[key] = state.attrs[key]
 		}
 	}
 
-	if (className != null) {
-		if (attrs.class != null) {
-			attrs.class = undefined
-			attrs.className = className
-		}
-
-		if (state.attrs.className != null) {
-			attrs.className = state.attrs.className + " " + className
-		}
+	if (state.attrs.className != null) {
+		attrs[classAttr] = state.attrs.className + (className != null ? " " + className : "")
+	} else if (className != null) {
+		attrs[classAttr] = className
```

Comment:
I've slightly changed your suggestion so that only `null` and `undefined` are treated as an absence of value. The other falsy values will either be stringified here or by `setAttribute()`.